File: android/sdk/src/main/java/com/viant/forgeandroid/ui/TreeEditorRenderer.kt
package com.viant.forgeandroid.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.Checkbox
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.viant.forgeandroid.runtime.ContainerDef
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.JsonUtil
import com.viant.forgeandroid.runtime.TreeEditorSpec
import com.viant.forgeandroid.runtime.WindowContext
import com.viant.forgeandroid.runtime.WorkflowPrimitiveRuntime

@Composable
internal fun TreeEditorRenderer(runtime: ForgeRuntime, window: WindowContext, container: ContainerDef) {
    val spec = container.treeEditor ?: return
    val dataSourceRef = spec.dataSourceRef ?: container.dataSourceRef ?: return
    val context = window.contextOrNull(dataSourceRef) ?: return
    val nodes by context.collection.flow.collectAsState(initial = context.collection.peek())
    var selected by remember(dataSourceRef) { mutableStateOf(emptySet<String>()) }
    var excluded by remember(dataSourceRef) { mutableStateOf(emptySet<String>()) }
    var expanded by remember(dataSourceRef) { mutableStateOf(emptySet<String>()) }
    var search by remember { mutableStateOf("") }

    LaunchedEffect(dataSourceRef) { if (context.dataSource.autoFetch != false) context.fetchCollection() }
    LaunchedEffect(nodes) {
        val hydration = hydrateTree(nodes, spec)
        if (selected.isEmpty()) selected = hydration.first
        if (excluded.isEmpty()) excluded = hydration.second
        if (expanded.isEmpty()) expanded = hydration.third
    }
    val rows = flattenTree(nodes, spec, expanded, search)
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        if (spec.searchable != false) OutlinedTextField(value = search, onValueChange = { search = it }, label = { Text("Search") }, modifier = Modifier.fillMaxWidth())
        if (rows.isEmpty()) Text(spec.emptyMessage ?: "No items")
        rows.forEach { row ->
            Row(modifier = Modifier.fillMaxWidth().padding(start = (row.depth * 16).dp), horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                if (row.hasChildren && spec.collapsible != false) Button(onClick = { expanded = if (row.id in expanded) expanded - row.id else expanded + row.id }) { Text(if (row.id in expanded) "−" else "+") }
                if (spec.selectionMode.equals("includeExclude", true)) {
                    val current = if (row.id in excluded) 2 else if (row.id in selected) 1 else 0
                    Row(modifier = Modifier.weight(1f), horizontalArrangement = Arrangement.spacedBy(4.dp)) {
                        listOf("None", "Include", "Exclude").forEachIndexed { index, label ->
                            Button(
                                enabled = current != index,
                                onClick = {
                                    selected = selected - row.id
                                    excluded = excluded - row.id
                                    if (index == 1) selected = selected + row.id
                                    if (index == 2) excluded = excluded + row.id
                                }
                            ) { Text(label) }
                        }
                    }
                } else {
                    Checkbox(
                        checked = row.id in selected,
                        onCheckedChange = { checked -> selected = WorkflowPrimitiveRuntime.toggleTreeSelection(nodes, selected, row.id, checked, spec) }
                    )
                    Text(row.label, modifier = Modifier.padding(top = 12.dp))
                }
            }
        }
        spec.mutation?.let { command ->
            MutationCommandButton(
                runtime, window, context, command,
                labelOverride = command.label ?: "Save selection",
                extras = mapOf("selectedIds" to selected.sorted(), "excludedIds" to excluded.sorted())
            )
        }
    }
}

private data class AndroidTreeRow(val id: String, val label: String, val depth: Int, val hasChildren: Boolean)

private fun flattenTree(nodes: List<Map<String, Any?>>, spec: TreeEditorSpec, expanded: Set<String>, search: String): List<AndroidTreeRow> {
    val result = mutableListOf<AndroidTreeRow>()
    fun visit(rows: List<Map<String, Any?>>, depth: Int) {
        rows.forEach { row ->
            val id = treeIdentity(row, spec) ?: return@forEach
            val label = row[spec.labelField ?: "label"]?.toString() ?: id
            val children = treeChildren(row, spec)
            if (search.isBlank() || label.contains(search, true)) result += AndroidTreeRow(id, label, depth, children.isNotEmpty())
            if (id in expanded || search.isNotBlank()) visit(children, depth + 1)
        }
    }
    visit(nodes, 0)
    return result
}

private fun hydrateTree(nodes: List<Map<String, Any?>>, spec: TreeEditorSpec): Triple<Set<String>, Set<String>, Set<String>> {
    val selected = mutableSetOf<String>(); val excluded = mutableSetOf<String>(); val expanded = mutableSetOf<String>()
    fun visit(rows: List<Map<String, Any?>>, depth: Int) {
        rows.forEach { row ->
            val id = treeIdentity(row, spec) ?: return@forEach
            if (spec.selectedField?.let { row[it] == true } == true) selected += id
            if (spec.excludedField?.let { row[it] == true } == true) excluded += id
            if (depth < (spec.defaultExpandedDepth ?: 0)) expanded += id
            visit(treeChildren(row, spec), depth + 1)
        }
    }
    visit(nodes, 0)
    return Triple(selected, excluded, expanded)
}

private fun treeIdentity(row: Map<String, Any?>, spec: TreeEditorSpec): String? = row[spec.identityField ?: "id"]?.toString()?.takeIf(String::isNotBlank)

private fun treeChildren(row: Map<String, Any?>, spec: TreeEditorSpec): List<Map<String, Any?>> =
    (row[spec.childrenField ?: "children"] as? List<*>)?.map(JsonUtil::asStringMap).orEmpty()

File: ios/Sources/ForgeIOSUI/TreeEditorRenderer.swift
import SwiftUI
import ForgeIOSRuntime

struct TreeEditorRenderer: View {
    let runtime: ForgeRuntime
    let window: WindowContext
    let container: ContainerDef

    @State private var nodes: [[String: JSONValue]] = []
    @State private var selected: Set<String> = []
    @State private var excluded: Set<String> = []
    @State private var expanded: Set<String> = []
    @State private var search = ""

    private var spec: TreeEditorSpec { container.treeEditor! }
    private var dataSourceRef: String { spec.dataSourceRef ?? container.dataSourceRef ?? "" }
    private var rows: [TreeEditorRow] { flattenTree(nodes, spec: spec, expanded: expanded, search: search) }

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            if spec.searchable != false {
                TextField("Search", text: $search).textFieldStyle(.roundedBorder)
            }
            if rows.isEmpty {
                Text(spec.emptyMessage ?? "No items").foregroundStyle(.secondary)
            } else {
                ForEach(rows) { row in
                    HStack(spacing: 6) {
                        Color.clear.frame(width: CGFloat(row.depth * 16), height: 1)
                        if row.hasChildren && spec.collapsible != false {
                            Button { toggleExpanded(row.id) } label: {
                                Image(systemName: expanded.contains(row.id) ? "chevron.down" : "chevron.right")
                            }.buttonStyle(.plain).accessibilityLabel(expanded.contains(row.id) ? "Collapse \(row.label)" : "Expand \(row.label)")
                        } else { Color.clear.frame(width: 12, height: 1) }
                        if spec.selectionMode?.lowercased() == "includeexclude" {
                            Picker(row.label, selection: includeExcludeBinding(row.id)) {
                                Text("None").tag(0); Text("Include").tag(1); Text("Exclude").tag(2)
                            }.pickerStyle(.segmented)
                        } else {
                            Toggle(row.label, isOn: selectedBinding(row.id))
                            #if os(macOS)
                                .toggleStyle(.checkbox)
                            #endif
                        }
                    }
                }
            }
            if let command = spec.mutation {
                MutationCommandButton(
                    runtime: runtime,
                    window: window,
                    sourceDataSourceRef: dataSourceRef,
                    command: command,
                    labelOverride: command.label ?? "Save selection",
                    extras: [
                        "selectedIds": .array(selected.sorted().map(JSONValue.string)),
                        "excludedIds": .array(excluded.sorted().map(JSONValue.string))
                    ]
                )
            }
        }
        .task(id: "\(window.windowID)#\(dataSourceRef)") { await observeNodes() }
        .accessibilityIdentifier("forge-tree-editor")
    }

    private func selectedBinding(_ key: String) -> Binding<Bool> {
        Binding(
            get: { selected.contains(key) },
            set: { selected = WorkflowPrimitiveRuntime.toggleTreeSelection(nodes: nodes, selected: selected, key: key, checked: $0, spec: spec) }
        )
    }

    private func includeExcludeBinding(_ key: String) -> Binding<Int> {
        Binding(
            get: { excluded.contains(key) ? 2 : selected.contains(key) ? 1 : 0 },
            set: { value in
                selected.remove(key); excluded.remove(key)
                if value == 1 { selected.insert(key) }
                if value == 2 { excluded.insert(key) }
            }
        )
    }

    private func toggleExpanded(_ key: String) { if expanded.contains(key) { expanded.remove(key) } else { expanded.insert(key) } }

    @MainActor
    private func observeNodes() async {
        nodes = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: dataSourceRef)
        if nodes.isEmpty { await runtime.refreshDataSourceCollection(windowID: window.windowID, dataSourceRef: dataSourceRef) }
        nodes = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: dataSourceRef)
        hydrateTreeState()
        let stream = await runtime.dataSourceCollectionUpdates(windowID: window.windowID, dataSourceRef: dataSourceRef)
        for await next in stream { nodes = next; hydrateTreeState() }
    }

    private func hydrateTreeState() {
        var initialExpanded: Set<String> = []
        var initialSelected: Set<String> = []
        var initialExcluded: Set<String> = []
        walkTree(nodes, spec: spec) { row, depth in
            guard let key = treeIdentity(row, spec: spec) else { return }
            if depth < (spec.defaultExpandedDepth ?? 0) { initialExpanded.insert(key) }
            if let field = spec.selectedField, row[field]?.boolValue == true { initialSelected.insert(key) }
            if let field = spec.excludedField, row[field]?.boolValue == true { initialExcluded.insert(key) }
        }
        if expanded.isEmpty { expanded = initialExpanded }
        if selected.isEmpty { selected = initialSelected }
        if excluded.isEmpty { excluded = initialExcluded }
    }
}

private struct TreeEditorRow: Identifiable {
    let id: String
    let label: String
    let depth: Int
    let hasChildren: Bool
}

private func flattenTree(_ nodes: [[String: JSONValue]], spec: TreeEditorSpec, expanded: Set<String>, search: String) -> [TreeEditorRow] {
    var result: [TreeEditorRow] = []
    func visit(_ rows: [[String: JSONValue]], depth: Int) {
        for row in rows {
            guard let id = treeIdentity(row, spec: spec) else { continue }
            let label = treeLabel(row, spec: spec)
            let children = treeChildren(row, spec: spec)
            if search.isEmpty || label.localizedCaseInsensitiveContains(search) {
                result.append(TreeEditorRow(id: id, label: label, depth: depth, hasChildren: !children.isEmpty))
            }
            if expanded.contains(id) || !search.isEmpty { visit(children, depth: depth + 1) }
        }
    }
    visit(nodes, depth: 0)
    return result
}

private func walkTree(_ nodes: [[String: JSONValue]], spec: TreeEditorSpec, visit: ([String: JSONValue], Int) -> Void, depth: Int = 0) {
    for node in nodes { visit(node, depth); walkTree(treeChildren(node, spec: spec), spec: spec, visit: visit, depth: depth + 1) }
}

private func treeChildren(_ row: [String: JSONValue], spec: TreeEditorSpec) -> [[String: JSONValue]] {
    row[spec.childrenField ?? "children"]?.arrayValue?.compactMap(\.objectValue) ?? []
}

private func treeIdentity(_ row: [String: JSONValue], spec: TreeEditorSpec) -> String? {
    let value = row[spec.identityField ?? "id"]
    switch value { case .string(let value): return value; case .number(let value): return value.rounded() == value ? String(Int(value)) : String(value); default: return nil }
}

private func treeLabel(_ row: [String: JSONValue], spec: TreeEditorSpec) -> String {
    row[spec.labelField ?? "label"]?.stringValue ?? treeIdentity(row, spec: spec) ?? "Item"
}
